File: app/src/main/java/com/luccaa53mini/S1Schedule.java
package com.luccaa53mini;

import android.content.Context;

/**
 * Represents the complete 7-day boiler schedule.
 *
 * Wire format: 84 bytes = 7 days × 12 bytes = 7 days × 3 slots × 4 bytes.
 * Each 4-byte slot: [on_min][on_hour][off_min][flags_off_hour]
 *   flags_off_hour: bit7 = enabled (1=slot active), bits6-0 = off_hour
 */
public class S1Schedule {

    public static final int DAYS  = 7;
    public static final int SLOTS = 3;

    /** [day 0-6][slot 0-2] */
    public final TimeSlot[][] slots = new TimeSlot[DAYS][SLOTS];

    public static final String[] DAY_NAMES =
            {"Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"};
    public static final String[] DAY_SHORT =
            {"Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"};

    public S1Schedule() {
        for (int d = 0; d < DAYS; d++)
            for (int s = 0; s < SLOTS; s++)
                slots[d][s] = new TimeSlot();
    }

    // ── Serialise to 84-byte wire format ────────────────────────────────────
    public byte[] toBytes() {
        byte[] out = new byte[84];
        for (int d = 0; d < DAYS; d++) {
            for (int s = 0; s < SLOTS; s++) {
                TimeSlot ts = slots[d][s];
                int base = d * 12 + s * 4;
                if (ts.enabled) {
                    out[base]     = (byte) (ts.offMinute & 0x3F);
                    out[base + 1] = (byte) (ts.offHour    & 0x1F);
                    out[base + 2] = (byte) (ts.onMinute & 0x3F);
                    out[base + 3] = (byte) (0x80 | (ts.onHour & 0x7F));
                } else {
                    out[base] = out[base+1] = out[base+2] = out[base+3] = 0;
                }
            }
        }
        return out;
    }

    // ── Deserialise from 84-byte wire format (or shorter partial) ───────────
    public static S1Schedule fromBytes(byte[] raw) {
        S1Schedule sched = new S1Schedule();
        if (raw == null) return sched;
        for (int d = 0; d < DAYS; d++) {
            for (int s = 0; s < SLOTS; s++) {
                int base = d * 12 + s * 4;
                if (base + 3 >= raw.length) return sched;
                int b0 = raw[base]     & 0xFF;
                int b1 = raw[base + 1] & 0xFF;
                int b2 = raw[base + 2] & 0xFF;
                int b3 = raw[base + 3] & 0xFF;
                TimeSlot ts = sched.slots[d][s];
                ts.enabled   = (b3 & 0x80) != 0;
                ts.offMinute = b0 & 0x3F;
                ts.offHour   = b1 & 0x1F;
                ts.onMinute  = b2 & 0x3F;
                ts.onHour    = b3 & 0x7F;
            }
        }
        return sched;
    }

    // ── TimeSlot ─────────────────────────────────────────────────────────────
    public static class TimeSlot {
        public boolean enabled   = false;
        public int     onHour    = 7;
        public int     onMinute  = 0;
        public int     offHour   = 8;
        public int     offMinute = 0;

        public String onTime(Context context) {
            java.util.Calendar cal = java.util.Calendar.getInstance();
            cal.set(java.util.Calendar.HOUR_OF_DAY, onHour);
            cal.set(java.util.Calendar.MINUTE, onMinute);
            return android.text.format.DateFormat.getTimeFormat(context).format(cal.getTime());
        }

        public String offTime(Context context) {
            java.util.Calendar cal = java.util.Calendar.getInstance();
            cal.set(java.util.Calendar.HOUR_OF_DAY, offHour);
            cal.set(java.util.Calendar.MINUTE, offMinute);
            return android.text.format.DateFormat.getTimeFormat(context).format(cal.getTime());
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            TimeSlot timeSlot = (TimeSlot) o;
            return enabled == timeSlot.enabled &&
                    onHour == timeSlot.onHour &&
                    onMinute == timeSlot.onMinute &&
                    offHour == timeSlot.offHour &&
                    offMinute == timeSlot.offMinute;
        }

        @Override
        public int hashCode() {
            return java.util.Objects.hash(enabled, onHour, onMinute, offHour, offMinute);
        }

        /** Returns true if on-time == off-time (invalid slot) */
        public boolean isConflict() {
            return enabled && onHour == offHour && onMinute == offMinute;
        }
    }
}

File: iOS/Lucca A53 mini/S1Schedule.swift
import Foundation
import Combine

// MARK: - S1Schedule
// Wire format: 84 bytes = 7 days × 12 bytes = 7 days × 3 slots × 4 bytes
// Each 4-byte slot: [offMin][offHour][onMin][flags_onHour]
//   flags_onHour: bit7 = enabled, bits6-0 = onHour

struct S1Schedule {

    static let DAYS  = 7
    static let SLOTS = 3

    static let dayNames: [String] = ["Sunday","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday"]
    static let dayShort: [String] = ["Sun","Mon","Tue","Wed","Thu","Fri","Sat"]

    /// [day 0-6][slot 0-2]
    var slots: [[TimeSlot]]

    // BUG 10 FIX: explicit nested-loop init avoids any risk of COW aliasing
    // when inner arrays happen to share the same backing store before mutation.
    init() {
        slots = (0..<S1Schedule.DAYS).map { _ in
            (0..<S1Schedule.SLOTS).map { _ in TimeSlot() }
        }
    }

    // MARK: - Serialize to 84-byte wire format
    func toBytes() -> [UInt8] {
        var out = [UInt8](repeating: 0, count: 84)
        for d in 0..<S1Schedule.DAYS {
            for s in 0..<S1Schedule.SLOTS {
                let ts = slots[d][s]
                guard ts.enabled else { continue }
                let base = d * 12 + s * 4
                out[base]     = UInt8(ts.offMinute & 0x3F)
                out[base + 1] = UInt8(ts.offHour   & 0x1F)
                out[base + 2] = UInt8(ts.onMinute  & 0x3F)
                out[base + 3] = UInt8(0x80 | (ts.onHour & 0x7F))
            }
        }
        return out
    }

    // MARK: - Deserialize from 84-byte wire format
    static func fromBytes(_ raw: [UInt8]) -> S1Schedule {
        var sched = S1Schedule()
        guard raw.count >= 84 else { return sched }
        for d in 0..<DAYS {
            for s in 0..<SLOTS {
                let base = d * 12 + s * 4
                let b0 = Int(raw[base])
                let b1 = Int(raw[base + 1])
                let b2 = Int(raw[base + 2])
                let b3 = Int(raw[base + 3])
                sched.slots[d][s].enabled   = (b3 & 0x80) != 0
                sched.slots[d][s].offMinute = b0 & 0x3F
                sched.slots[d][s].offHour   = b1 & 0x1F
                sched.slots[d][s].onMinute  = b2 & 0x3F
                sched.slots[d][s].onHour    = b3 & 0x7F
            }
        }
        return sched
    }

    // MARK: - TimeSlot
    struct TimeSlot: Equatable {
        var enabled:   Bool = false
        var onHour:    Int  = 7
        var onMinute:  Int  = 0
        var offHour:   Int  = 8
        var offMinute: Int  = 0

        var onTimeString:  String { formatted(hour: onHour,  minute: onMinute) }
        var offTimeString: String { formatted(hour: offHour, minute: offMinute) }

        private func formatted(hour: Int, minute: Int) -> String {
            var c = DateComponents(); c.hour = hour; c.minute = minute
            let date = Calendar.current.date(from: c) ?? Date()
            let f = DateFormatter(); f.timeStyle = .short; f.dateStyle = .none
            return f.string(from: date)
        }
    }
}

// MARK: - UiEntry
// Aggregates slots that share the same ON/OFF time across multiple days into one card.
// Mirrors the Android UiEntry.
class UiEntry: ObservableObject, Identifiable, Equatable {
    let id = UUID()
    @Published var onH:  Int = 7
    @Published var onM:  Int = 0
    @Published var offH: Int = 8
    @Published var offM: Int = 0
    @Published var days: [Bool] = Array(repeating: false, count: 7)  // Mon=0..Sun=6

    init() {}
    init(onH: Int, onM: Int, offH: Int, offM: Int, days: [Bool]) {
        self.onH = onH; self.onM = onM
        self.offH = offH; self.offM = offM
        self.days = days
    }

    var onTimeString:  String { formatTime(onH,  onM)  }
    var offTimeString: String { formatTime(offH, offM) }

    private func formatTime(_ h: Int, _ m: Int) -> String {
        var c = DateComponents(); c.hour = h; c.minute = m
        let date = Calendar.current.date(from: c) ?? Date()
        let f = DateFormatter(); f.timeStyle = .short; f.dateStyle = .none
        return f.string(from: date)
    }

    static func == (lhs: UiEntry, rhs: UiEntry) -> Bool { lhs.id == rhs.id }
}

// MARK: - Schedule Conversion Helpers
// UI day order : Sun=0, Mon=1, Tue=2, Wed=3, Thu=4, Fri=5, Sat=6
// HW day order : Sun=0, Mon=1, Tue=2, Wed=3, Thu=4, Fri=5, Sat=6  (Java Calendar)
// hwToUi[hwDay] -> uiDay index
// uiToHw[uiDay] -> hwDay index

extension S1Schedule {

    static func loadUiEntries(from hw: S1Schedule) -> [UiEntry] {
        var entries: [UiEntry] = []
        var map: [String: UiEntry] = [:]
        let hwToUi = [1, 2, 3, 4, 5, 6, 0]   // Sun(0)->0, Mon(1)->1, … Sat(6)->6

        for hwDay in 0..<DAYS {
            let uiDay = hwToUi[hwDay]
            for s in 0..<SLOTS {
                let ts = hw.slots[hwDay][s]
                guard ts.enabled else { continue }
                let key = String(format: "%02d:%02d-%02d:%02d",
                                 ts.onHour, ts.onMinute, ts.offHour, ts.offMinute)
                if let entry = map[key] {
                    entry.days[uiDay] = true
                } else {
                    let entry = UiEntry(onH: ts.onHour, onM: ts.onMinute,
                                        offH: ts.offHour, offM: ts.offMinute,
                                        days: Array(repeating: false, count: 7))
                    entry.days[uiDay] = true
                    map[key] = entry
                    entries.append(entry)
                }
            }
        }
        return entries
    }

    static func convertUiToHardware(_ entries: [UiEntry]) -> S1Schedule {
        var hw = S1Schedule()
        var slotCounts = [Int](repeating: 0, count: 7)
        let uiToHw = [6, 0, 1, 2, 3, 4, 5]   // Sun(0)->0, Mon(1)->1, … Sat(6)->6

        for e in entries {
            for uiDay in 0..<7 {
                guard e.days[uiDay] else { continue }
                let hwDay = uiToHw[uiDay]
                let slotIdx = slotCounts[hwDay]
                guard slotIdx < SLOTS else { continue }
                hw.slots[hwDay][slotIdx].enabled   = true
                hw.slots[hwDay][slotIdx].onHour    = e.onH
                hw.slots[hwDay][slotIdx].onMinute  = e.onM
                hw.slots[hwDay][slotIdx].offHour   = e.offH
                hw.slots[hwDay][slotIdx].offMinute = e.offM
                slotCounts[hwDay] += 1
            }
        }
        return hw
    }
}
